### PR TITLE
Improve privacy for hidden battles

### LINKED ISSUE #6950: User privacy enhancements
### Add a `/nofeedback` command

> I've been thinking about some sort of command like
> /nofeedback /hidenext so it doesn't print a response
> hence /nofeedback as a command that runs the next command, but hides all sendReply responses

### Add sticky checkboxes to the Find Battle/Challenge UIs which will cause the battle to be hidden

`<abbr title="You can still invite spectators by giving them the URL or using the /invite command" class="checkbox">Ban spectators</abbr>`

> I would put it between the team selector and "Battle!" button & above challenge/cancel and below the team selector

The state of the Ban Spectators checkbox can be stored in local storage to ensure it is sticky.

### Support renaming rooms

> apparently the rename functionality just adds a broadcast linking to the new room
> but we do also have protocol infrastructure for actually renaming a room
> it's used by room aliases to redirect to the actual roomid

### Passwords

When a room is made invite only (either from the beginning or mid battle), kick out all guests and randomize the URL

- the new room id (and thus URL) generated by the sim server should follow the same pattern as replays which were made private: https://replay.pokemonshowdown.com/gen7randombattle-956127160-fz0wmd2n1wu7gorbay9jtvqxpc3tkl4pw. ie/ normal room id plus a random character string that ends in 'pw' (for 'password')
- if a replay is saved of the invite only room, the password generated by the sim server is to be used instead of the password generated by the replay server
- the password generation on the replay server is still needed in case a replay is made private after the fact, though the password generated may be different than the password generated by the sim server (though will follow the same format)

### PR DESCRIPTION
**Pull request by AnnikaCodes**
Incorporates [these changes](https://github.com/smogon/pokemon-showdown/issues/6950#issuecomment-653822792) and also kicks guests out of battles while hiding them.

Also, I'm pretty sure this fixes #6950.

### CODE REVIEW

File: server/rooms.ts [hunk 1]
```diff
@@ -1756,11 +1783,11 @@ export const Rooms = {
 
 		if (privacySetter.size) {
 			if (battle.forcePublic) {
-				room.settings.isPrivate = false;
+				void room.makePublic();
```
Comment (scheibo) [inline server/rooms.ts:1786]: why are we not propagating the promise here and below?

File: server/modlog.ts [hunk 2]
```diff
@@ -171,7 +171,9 @@ export class Modlog {
 	async rename(oldID: ModlogID, newID: ModlogID) {
 		const streamExists = this.streams.has(oldID);
 		if (streamExists) await this.destroy(oldID);
-		await FS(`${this.logPath}/modlog_${oldID}.txt`).rename(`${this.logPath}/modlog_${newID}.txt`);
+		if (!this.getSharedID(oldID)) {
```
Comment (scheibo) [inline server/modlog.ts:174]: is this change related? should it be in a separate PR?

File: server/rooms.ts (hunk 1 above)
Comment (AnnikaCodes) [reply server/rooms.ts:1786]: Because propagating clutters up a lot of unrelated code with `async`s (including a `setTimeout` in ladder that would look a lot less readable). It's definitely possible to propagate it but I don't forsee it causing problems.

File: server/modlog.ts (hunk 2 above)
Comment (AnnikaCodes) [reply server/modlog.ts:174]: No, it's necessary; otherwise, trying to rename the battle crashes PS!

File: server/rooms.ts (hunk 1 above)
Comment (scheibo) [reply server/rooms.ts:1786]: I strongly disagree with ever `void`-ing promises (at the very least they need to be propagated to the top level and cases of ignoring it at the top level should be accompanied with a comment justifying it) and I really which PS didn't normalize this and teach developers it is OK to sacrifice correctness at the alter of 'avoiding clutter', but I've lost that battle.

Comment (Zarel) [reply server/rooms.ts:1786]: Unnecessarily propagating promises also makes the code less parallel and ultimately slower. It looks like the async-ness here comes from `rename`, which I'm pretty sure doesn't need to be propagated outside of the relevant roomlog function at all (the relevant roomlog functions needs to be able to handle the rename without race conditions whether or not the Promise is propagated). But I suppose you're not interested in having this argument.

Comment (scheibo) [reply server/rooms.ts:1786]: If you believe changing the argument in favor from 'avoiding clutter' to 'performance reasons' is going to convince me that making something untestable and incorrect is going to satisfy me then yes, I'm not interested in having that argument.

Feel free to ignore this, of course. I am never going to approve/merge code that doesn't propagate promises, but you do you.

Comment (AnnikaCodes) [reply server/rooms.ts:1786]: (I'm happy to propagate it or not once you two figure this out - I was just going off PS' existing code style choices)

Comment (Zarel) [reply server/rooms.ts:1786]: It doesn't seem like I'm going to understand pre's perspective; he's explained it to me a few times and it's never clicked for me. Let's just put the `void` inside `this.rename` (change the last `return` to `void`).

File: server/modlog.ts [hunk 3]
```diff
@@ -171,7 +171,9 @@ export class Modlog {
...
+			await FS(`${this.logPath}/modlog_${oldID}.txt`).rename(`${this.logPath}/modlog_${newID}.txt`);
+		}
 		if (streamExists) this.initialize(newID);
```
Comment (Zarel) [inline server/modlog.ts:177]: So, the problem is that this function races with `this.write`.

Look at `this.write`:

```
	write(roomid: ModlogID, message: string, overrideID?: string) {
		const stream = this.streams.get(roomid);
		if (!stream) throw new Error(`Attempted to write to an uninitialized modlog stream for the room '${roomid}'`);
		void stream.write(`[${new Date().toJSON()}] (${overrideID || roomid}) ${message}\n`);
	}
```

It will crash if it's called during either of the two `await`s in this rename function.

To be fair, this isn't a problem your PR introduces, and propagating the Promise will not fix this problem at all... but it does need to be fixed at some point.

Comment (Zarel) [reply server/modlog.ts:177]: If renames are atomic (which I believe they nearly always are), we could initialize the new stream _immediately_ instead of waiting. So `void` instead of `await` in every call here. I believe that's sufficient to solve the race condition (except in obscure situations where the rename isn't atomic).

If we can't assume that the rename is atomic or want to do this the "right" way, you would want to store a promise for the stream that `write` can use to wait until it can write for real.

Comment (AnnikaCodes) [reply server/modlog.ts:177]: > it does need to be fixed at some point.

We're migrating modlogs to SQLite, which should solve the race condition.

**Comment by Zarel** [comment]
Does this require a server restart?

**Comment by AnnikaCodes** [comment]
Oh, yeah, it does. I always forget to assign the label.

**Review by Zarel — APPROVED** [review]
This is great! I think it's ready for the next restart.

**Comment by Zarel** [comment]
oh no conflict

**Comment by Zarel** [comment]
(I'll deal with it)
